### ISSUE TITLE
Priority & Preemption should be restricted-by-default

### ISSUE THREAD

**Issue by jberkus** [kubernetes/kubernetes#65557]
/kind bug
/area security
/sig scheduling
/priority important-soon

Currently, an admin who has Priority & Preemption enabled (as they are by default) and has  ResourceQuotaScopeSelectors disabled (as they are by default) has no way to restrict regular users of Kubernetes from creating as many system-node-critical and system-cluster-critical pods as they care to.  This can result in creating a denial-of-service if users abuse priorities in competition with each other.

By 1.12, we could have ResourceQuotaScopeSelectors enabled by default and add a default restriction on the use of system-* priorities.  However, that kind of leaves 1.11 users in the lurch.  What can we do to restrict the ability of regular users to abuse system-* priorities that can be added to 1.11.1?

/assign @bsalamat 
/cc @derekwaynecarr @vikaschoudhary16 @liggitt

**Comment by liggitt** [comment]
That's also been true of pods in general since 1.0 (unlimited unless a resource quota is in place). @derekwaynecarr can chime in with the intended approach for making it easy for an admin who wants to limit this to indicate that cluster-wide

**Comment by jberkus** [comment]
AFAIK, though, it hasn't been previously possible for two misbehaving users in a preemption race to accidentally shut down kubedns, or similar.  That's what I'd like to see some tools to protect against.

**Comment by bsalamat** [comment]
The risks here is that users may maliciously or inadvertently create a large number of pods at system priority levels that would cause some of the system critical pods running at system-cluster-critical priority to get preempted.

Our plan to prevent the risk was to introduce [Pod Priority support in ResourceQuota ](https://github.com/kubernetes/kubernetes/pull/57963) when moving Pod Priority and Preemption to Beta. The feature was added to ResourceQuota, but unfortunately, it is gated by default in 1.11. Kubernetes users can still enable scopeSelector feature of ResourceQuota and use it to remove the potential DoS risk.

If users do not want to enable the feature, an alternative approach is to set regular ResourceQuota (without the scopeSelector) to limit users from creating unlimited number of pods in a cluster. This approach is not as effective as using scopeSelector, but still prevents users from creating a large number of pods at system priority that would fill up a cluster and could cause preemption of some system components.

A third alternative is to run all the system components with system-node-critical priority. Many of the system components already run with system-node-critical priority and therefore are not preempted, but there are some system components, namely, kube-dns, that run with system-cluster-critical. Changing their priority to system-node-critical can prevent them from getting preempted. This approach is probably the least effective approach, as other user workloads are still susceptible to preemption by a large number of pods created at system priority levels.

**Comment by bsalamat** [comment]
What I would suggest to do for 1.11.1 is to change our admission controller to allow the use of `system-node-critical` and `system-cluster-critical` priorities only in `kube-system` namespace. Then our current (1.11) ResourceQuota can be used to limit users from creating pods in `kube-system` namespace. With that users won't be able to use the two system PriorityClasses and the risk of abusing them would be mitigated.

**Comment by smarterclayton** [comment]
The feature was gated because it was a new field on a GA API and we do not promote new fields directly to beta on GA APIs.  I just want to be clear that the bar for new APIs doesn't get lowered because another thing is going to beta.  

The admission controller suggestion is reasonable, although derek's is probably more appropriate.

**Comment by davidopp** [comment]
The admission controller idea is great.

We should update the release note to reflect the change.

**Comment by bsalamat** [comment]
I am fine with expanding the permitted namespaces from `kube-system` to a configurable set of namespaces in our priority admission controller.

**Comment by liggitt** [comment]
> I would like to add that the default permitted namespace of the admission controller should be kube-system. If a configuration is provided, it will override the default and may expand the namespaces.

agreed. that maintains parity with the pre-API alpha critical pod annotation use

**Comment by aveshagarwal** [comment]
So for final plan, as agreed, @bsalamat will send an initial PR to restrict system priority classes to kube-system namespace for 1.11.1. After that I will work on a follow-up PR to make namespace restriction configurable in priority admission plugin.

**Comment by jberkus** [comment]
@bsalamat please let @foxish know when you'll have this fix, we'll be scheduling 1.11.1 based on it.  Thanks!

**Comment by bsalamat** [comment]
@jberkus @foxish I just sent the above PR against the master branch. I will cherrypick it into 1.11 branch once it is reviewed.

**Comment by jberkus** [comment]
Cool, thanks!

**Comment by derekwaynecarr** [comment]
@jberkus to clarify, please hold 1.11.1 on both PRs
https://github.com/kubernetes/kubernetes/issues/65557#issuecomment-401067454

**Comment by luxas** [comment]
Does the proposed solution work when the `Priority` admission controller is disabled (as it is by default)?
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/options/plugins.go#L131-L139
Should e.g. kubeadm start doing `--enable-admission-plugins=Priority` in v1.11.1 to fix this until it's enabled by default (or similar) in v1.12?

**Comment by aveshagarwal** [comment]
> Does the proposed solution work when the Priority admission controller is disabled (as it is by default)?

No. As P&P is enabled by default, my understanding is that Priority admission plugin should be enabled by default too.

**Comment by liggitt** [comment]
> No. As P&P is enabled by default, my understanding is that Priority admission plugin should be enabled by default too.

Any cluster that specifies `--admission-control` determines for itself whether the Priority admission plugin is enabled. Any cluster that is using a value from previous releases is unlikely to have that admission plugin enabled.

**Comment by aveshagarwal** [comment]
Here is a follow up PR: https://github.com/kubernetes/kubernetes/pull/65643

**Comment by luxas** [comment]
> As P&P is enabled by default

Hmm, the admission controller isn't: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/options/plugins.go#L131-L139.
Expected or not?

In this case, kubeadm doesn't use `--admission-control` but instead relies on the default list.
My question is, _would you recommend that_ the P&P admission controller is enabled in v1.11?

**Comment by bsalamat** [comment]
Priority admission controller must be enabled by default in 1.11. My understanding was that it was enabled even in 1.10 (I am on mobile and cannot check right now), but the feature gate flag was preventing the admission controller from resolving priority values. In 1.11, the feature gate is enabled and the admission controller resolves priorities.

Regardless, if the admission controller is disabled, priority classes will no be resolved to the integer values. In other words, no pod will have any priority (or you can say all pods will have the same priority). So, the issue that we are discussing here won't exist.

**Comment by bsalamat** [comment]
> If the admission plugin is disabled, can't any integer priority value be set?

Good point, users would be able to set priority values manually in that case.

**Comment by bsalamat** [comment]
This is the cherrypick fo the PR into 1.11: https://github.com/kubernetes/kubernetes/pull/65717

**Comment by bsalamat** [comment]
> Hmm, the admission controller isn't: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/options/plugins.go#L131-L139.
Expected or not?

My understanding is that, these are plugins which are disabled by default and Priority is not among them.

**Comment by liggitt** [comment]
Any plugin not in the `defaultOnPlugins` list is disabled by default

**Comment by luxas** [comment]
Which means we probably have a bit broader problem wrt this as it's not enabled by default. Should we do that in v1.11.1 as well if it always was the intention in order to not let the user set any priority value (as the feature gate is enabled)?

**Comment by liggitt** [comment]
> Which means we probably have a bit broader problem wrt this as it's not enabled by default. Should we do that in v1.11.1 as well if it always was the intention in order to not let the user set any priority value (as the feature gate is enabled)?

Yes. https://github.com/kubernetes/kubernetes/pull/65722

**Comment by bsalamat** [comment]
> Any plugin not in the defaultOnPlugins list is disabled by default

I have not been able to find any list called `defaultOnPlugins`. Could you please send a link?

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/blob/92b81114f43f3ca74988194406957a5d1ffd1c5d/pkg/kubeapiserver/options/plugins.go#L130-L140

**Comment by bsalamat** [comment]
Thanks! The function name confused me.

**Comment by liggitt** [comment]
addresses by https://github.com/kubernetes/kubernetes/pull/65768 and https://github.com/kubernetes/kubernetes/pull/65717

/close

**Comment by ravisantoshgudimetla** [comment]
@liggit, I think we should let https://github.com/kubernetes/kubernetes/pull/65643 merge before closing this issue.

**Comment by liggitt** [comment]
the critical issue was protecting the feature. #65768 and #65717 do that effectively. if we want configuration to let it be re-expanded in a controlled way, that's fine, but the critical issue is resolved

**Comment by ravisantoshgudimetla** [comment]
Ok, I have created another issue to track the enhancement.
